### ISSUE TITLE
[Feature Request] Schema state import/export

### ISSUE THREAD

**Issue by shousper** [typeorm/typeorm#664]
Is there any plans to add a "schema state" file similar to the `schema.rb` file in RoR? I believe adding something like a `schema.json` that is generated and committed to source control could be very useful. Developer teams working in branches that both generate migrations and merge to mainline will not currently get any conflicts, but the last to merge will get a failed migration. If the `schema.json` is committed, this should generate conflicts that prevent such a scenario.

Thoughts?

**Comment by pleerock** [comment]
Sorry I'm not familiar with RoR and how this schema stuff works in it. Can you please provide a more detailed proposal how it should work in typeorm

**Comment by shousper** [comment]
Sure, I'll start with the problem I'm trying to solve.

Using a wildcard pattern like `/migration/*.js` in the migration configuration is perfect for preventing migrations getting missed and never executed and the CLI generator provides the boilerplate for new migrations. All of this works great. My team work in branches and often generate new migrations in each.

The problem we have is the new migrations don't conflict upon merge and instead fail upon deploy as the migrations run in an unexpected order. The migrations then require manual reordering of their sequence by renaming the migration files. This corrects the failing migration and is _definitely what should happen to prevent migrations running out of order_ (i.e. what you're already doing 👍)

What would be great is if that failure could happen before we merge. This is where I believe a `schema.json` export and verification can purposefully cause merge conflicts and integration test failures before code is merged. I would strongly recommend reading the [Active Record Migrations](http://guides.rubyonrails.org/active_record_migrations.html) page to better understand how they use the `schema.rb` file. Sections 6.1 and 6.3 are probably the most relevant right now though:

_From section 6.1_

> Migrations, mighty as they may be, are not the authoritative source for your database schema. That role falls to either db/schema.rb or an SQL file which Active Record generates by examining the database. They are not designed to be edited, they just represent the current state of the database.
>
> There is no need (and it is error prone) to deploy a new instance of an app by replaying the entire migration history. It is much simpler and faster to just load into the database a description of the current schema.

_From section 6.3_

> db/schema.rb contains the current version number of the database. This ensures conflicts are going to happen in the case of a merge where both branches touched the schema. When that happens, solve conflicts manually, keeping the highest version number of the two.

Basically we need that version number conflict, but we should also have an "authoritative source" for the schema. The schema definition can also be used to verify migrations have all run successfully. This could be useful for integration tests where the test database is migrated and verified before hand.

I hope this helps clarify the idea, sorry its a little long winded. Please let me know what else I can do to help.

**Comment by shousper** [comment]
@pleerock I see you tagged this "needs more info". Anything else I can do to help? :)

**Comment by pleerock** [comment]
This issue needs more of my time for deeper investigation. I would like to do something better then generating `schema.json` file. I don't like this approach, would be great to find some alternatives.

**Comment by pleerock** [comment]
Closing in favour of #664 @shousper would be great if you share your thoughts in discussion there.

**Comment by shousper** [comment]
@pleerock in favour of which issue? 😉

**Comment by pleerock** [comment]
ah sorry in favour of #1315

**Comment by shusson** [comment]
I don't think #1315 is an alternative to having a schema file. Right now if I would like to know if a PR has changed anything in the db, I have to look through code. Ideally I want to diff a schema file. 

Our planned workaround is to use pg_dump, but it would be nice if this was baked into typeorm.
